File: src/agents/maintenance/redundant-tests.ts
import type { TestRegistryItem, MaintenanceFinding } from '../shared/types';

function normalizeTitle(title: string) {
  return title
    .replace(/@[a-zA-Z0-9-_:]+/g, '')
    .toLowerCase()
    .replace(/\s+/g, ' ')
    .trim();
}

export async function findRedundantTests(
  registry: TestRegistryItem[]
): Promise<MaintenanceFinding[]> {
  const findings: MaintenanceFinding[] = [];

  const byFeature = new Map<string, TestRegistryItem[]>();

  for (const test of registry) {
    const existing = byFeature.get(test.feature) ?? [];
    existing.push(test);
    byFeature.set(test.feature, existing);
  }

  for (const [feature, tests] of byFeature.entries()) {
    const seen = new Map<string, TestRegistryItem[]>();

    for (const test of tests) {
      const normalized = normalizeTitle(test.title);
      const group = seen.get(normalized) ?? [];
      group.push(test);
      seen.set(normalized, group);
    }

    for (const group of seen.values()) {
      if (group.length > 1) {
        findings.push({
          type: 'redundant_test',
          severity: 'low',
          message: `Possible duplicate tests found in feature ${feature}`,
          payload: {
            feature,
            tests: group.map((t) => ({
              id: t.id,
              file: t.file,
              title: t.title
            }))
          }
        });
      }
    }
  }

  return findings;
}

File: python_agents/agents/maintenance/redundant_tests.py
from __future__ import annotations

import re

from ..shared.models import Finding, TestRegistryItem


TAG_PATTERN = re.compile(r"@[a-zA-Z0-9-_:]+")


def find_redundant_tests(registry: list[TestRegistryItem]) -> list[Finding]:
    findings: list[Finding] = []
    by_feature: dict[str, list[TestRegistryItem]] = {}

    for test in registry:
        by_feature.setdefault(test.feature, []).append(test)

    for feature, tests in by_feature.items():
        seen: dict[str, list[TestRegistryItem]] = {}
        for test in tests:
            normalized = _normalize_title(test.title)
            seen.setdefault(normalized, []).append(test)

        for group in seen.values():
            if len(group) > 1:
                findings.append(Finding(
                    type="redundant_test",
                    severity="low",
                    message=f"Possible duplicate tests found in feature {feature}",
                    payload={
                        "feature": feature,
                        "tests": [
                            {
                                "id": item.id,
                                "file": item.file,
                                "title": item.title,
                            }
                            for item in group
                        ],
                    },
                ))

    return findings


def _normalize_title(title: str) -> str:
    without_tags = TAG_PATTERN.sub("", title)
    return re.sub(r"\s+", " ", without_tags).lower().strip()
